vignettes/TechnicalDocmentation.Rmd

```diff
@@ -0,0 +1,636 @@
+---
+title: "Technical Documentation"
+header-includes:
+output: html_document
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = FALSE,message = FALSE)
+```
+
+This is internal package documentation describing the vectorized formulation of the Convex Clustering problem, CARP and CBASS.
+
+The goal is describe how CARP and CBASS make their updates at each iteration, and also describe some of the auxillary functions
+which are used to accomplish this.
+
+## Original Problem
+
+The Convex Clustering problem is given by
+
+$$
+\min_{\boldsymbol{U}} \frac{1}{2} \| \boldsymbol{X} - \boldsymbol{U} \|_F^2 + \lambda \sum_{(i,j) \in E} w_{ij} \| \boldsymbol{u}_{i} - \boldsymbol{u}_{j} \|_2
+$$
+here
+
+* $\boldsymbol{X} \in \mathbb{R}^{p \times n}$ is our data matrix consisting of $n$ observations in $p$ dimensions
+
+* $\lambda > 0$ a reguarlization parameter
+
+* $E$ our edge set of size $|E|$
+
+* $w_{ij} > 0$ a set of given weights; we have $|E|$ of these
+
+* $\boldsymbol{U} \in \mathbb{R}^{p \times n}$ is our estimand; we denote it's 
+$i$th column (the $i$ the observation) as $\boldsymbol{u}_{i} in \mathbb{R}^p$
+
+
+## Vectorized Problem
+
+We rewrite in terms of vectors. Define 
+
+* $\boldsymbol{x} \equiv \textrm{vec}(\boldsymbol{X}) \in \mathbb{R}^{np}$
+
+* $\boldsymbol{u} \equiv \textrm{vec}(\boldsymbol{u}) \in \mathbb{R}^{np}$
+
+
+
+The loss can be rewritten as $\frac{1}{2}\|\boldsymbol{x} - \boldsymbol{u}\|_2^2$.
+To rewrite the regularization term we need to compute $\boldsymbol{u}_i - \boldsymbol{u}_j$
+in term of $\boldsymbol{u}$.
+
+Define
+
+* $\boldsymbol{d}_{ij} \in \mathbb{R}^n$ be a vector of all $0$'s, except with a $1$ in 
+the $i$th position and $-1$ in the $j$th position.
+
+* $\boldsymbol{D}_{ij} \equiv (\boldsymbol{d}_{ij}^T \otimes \boldsymbol{I}_p)$
```

Comment:
Does working on `X` instead of `t(X)` change the order of Kronecker product? It's much faster to calculate `(I \otimes A)^{-1}b` than `(A \otimes I)^{-1}b` so it could be worth it to do any reformulation that lets us flip the product.